src/expiry.1
.\" -*- nroff -*-
.\" Copyright (C) 2003 Thorsten Kukuk
.\" Author: Thorsten Kukuk <kukuk@suse.de>
.\"
.\" This program is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License version 2 as
.\" published by the Free Software Foundation.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with this program; if not, write to the Free Software Foundation,
.\" Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
.\"
.TH expiry 1 "March 2003" "pwdutils"
.SH NAME
expiry \- check password expiration and enforce password change
.SH SYNOPSIS
.TP 5
\fBexpiry\fR [\fB-f\fR]
.SH DESCRIPTION
\fBexpiry\fR checks the current password expiration and
prints a warning if the password is expiring or enforces a change
when required.
.SH OPTIONS
.TP
.B "\-f"
The caller is forced to change the password if it is expired
but inactive.
.SH FILES
passwd \- user account information
.br
shadow \- shadow user account information
.SH SEE ALSO
.BR passwd (1),
.BR passwd (5)
.SH AUTHOR
Thorsten Kukuk <kukuk@suse.de>


NEWS
pwdutils NEWS -- history of user-visible changes.

Copyright (C) 2002-2012 Thorsten Kukuk

Please send bug reports, questions and suggestions to <kukuk@thkukuk.de>.

Version 3.2.19
* Fix passwd -S

Version 3.2.18
* Fix changing of permissions if uid/gid of user will be modified

Version 3.2.17
* Fix useradd.local

Version 3.2.16
* Make sha512 new default password hash

Version 3.2.15
* newgrp: honour dynamically assigned groups

Version 3.2.14
* pwck: check +/- entries, too
* grpck: check */- entries, too
* login.defs: remove LOGIN_RETRIES
 
Version 3.2.13
* Only copy regular files if usermod moves home directory

Version 3.2.12
* Make sure modified group files are written to disk
* Update translations for sv, fr

Version 3.2.11
* Update translations
* Add translation for nl

Version 3.2.10
* Fix encoding of polish translations
* Fix long options of grpck/pwck
 
Version 3.2.9
* Move UMASK from /etc/login.defs to /etc/default/useradd

Version 3.2.8
* Fix usermod manual page
* Add quiet option to pam_rpasswd.so

Version 3.2.7
* Fix possible endless loop in userdel

Version 3.2.6
* Fix rpasswd protocol error
* Update translations

Version 3.2.5
* Add translations for da
* pwconv: Remove temporary files in case of an error

Version 3.2.4
* Make sure changes are really written to disk

Version 3.2.3
* New translations
* Adjust documentation

Version 3.2.2
* Don't lock system accounts by default

Version 3.2.1
* Adjust for new libxcrypt

Version 3.2.0
* Better support for audit plugin
* --preferred-uid/gid option for useradd and groupadd

Version 3.1.4
* Update polish translation.
* Remove FTMP references.
* Better pwconv support for  NIS entries

Version 3.1.3
* Allow compiling without native language support.
* Update swedish translation.

Version 3.1.2
* Check if crontab is installed before calling it.
* Fix seg.fault in pwconv in error case.

Version 3.1.1
* Adjust PAM config files

Version 3.1.0
* gpasswd: Add -l and -u option to lock/unlock group passwords
* passwd: Add support for BioAPI
* Add LDAP support for RFC2307bis

Version 3.0.7
* passwd.1: correct manual page
* passwd: Allow changing passwords with PAM even if service is unknown

Version 3.0.6
* chfn: Fix syntax check for home phone number

Version 3.0.5
* Fix compilation error of PAM module with new Linux-PAM release
* Fix CHARACTER_CLASS in /etc/login.defs

Version 3.0.4
* Fix pam_rpasswd.so

Version 3.0.3
* usermod: Adjust sp_lstchg if -p option is used
* useradd: Preserve existing home directory
* chage: Fix off-by-one errors in output

Version 3.0.2
* Fix loading of logging plugins
* Unify options (use always --service)
* Add -a option to usermod for shadow compatibility

Version 3.0.1
* usermod: Don't remove old home directory if an error occured

Version 3.0
* Add --stdin option for passwd
* Fix useradd -D option
* Fix PAM/syslog interaction
* Fix lot of ignored failing syscalls
* Compile daemons with -fpie/-pie
* Add support for logging plugins
* Add plugin for LAuS support
* Add plugin for audit support
* Add plugin to log with syslog
* Use libnscd
* Add support for GnuTLS
* groupmod: Add options to remove/add single user from a group
* usermod: Add options to remove/add single user from a list of groups
* rpasswd/rpasswdd: Better SLP support

Version 2.6.4
* rpasswd/pam_rpasswd.so: Fix SLP URL parsing

Version 2.6.3
* newgrp: Add -c option to allow sg emulation

Version 2.6.2
* Fix compile problem if SELinux is disabled
* rpasswdd: Re-register SLP service after timeout

Version 2.6.1
* newgrp: Fix NGROUPS_MAX handling

Version 2.6
* Add -f compat flag to groupadd
* Add UTF8/local locale translation
* Make regex for allowed accounts/groups configurable
* Add optional pam_rpasswd module
* Add SLP support to rpasswd/rpasswdd
* Add pwck
* Add grpck
* Add newgrp
* Add /etc/default/passwd support

Version 2.5.1
* Install all manual pages

Version 2.5
* Adjust for libselinux v1.5
* Add pwconv/pwunconv
* Add grpunconv

Version 2.4
* Fix -G handling of usermod
* Fix LDAP support
* Add gpasswd, groupmod and usermod
* Copy always all extended attributes
* Add SELinux support for passwd, chsh and chfn
* Make sure that group, passwd and shadow are always created with
  correct security attributes
* groupadd -g handling fixed
* vipw/vigr added
* Lot of bug fixes
* Add manual pages
* Add groupadd and groupdel
* useradd and userdel can now create/delete LDAP accounts
* Make compile with OpenLDAP 2.0.x again
* Rewrite ldap functions to meet demands
* chpasswd can now change LDAP passwords
* useradd: Lot of new error checks and bug fixes
* chfn/chsh: Allow user to abort with Ctrl-C during data input
* Add test suite

Version 2.3
* Add chpasswd
* Add expiry
* Fix chage options
* rpasswd client now checks server certificate

Version 2.2
* Add "-D binddn" option, to allow modifying LDAP data with a special
  account.
* passwd and chage: Allow modifying of password expiry informations
  stored in a LDAP database.

Version 2.1
* Add --disable-ldap option to configure
* Fix typos in chage.1 manual page
* Rewrite passwd manual page and explain all options

Version 2.0
* Fix -d and -E option of chage
* Use port 774 as new default
* Make compile with glibc 2.3 and gcc 2.95/3.2/3.3

Version 1.99
* Add chage
* Allow modifying of passwd/shadow file not in /etc (only chage yet)

Version 1.98
* Add manual pages for rpasswd and rpasswdd

Version 1.97
* rpasswd now reads config file with server name and port number
* Security fix for rpasswdd: don't allow remote changing of root
  password without authentification
* Add manual page for rpasswd.conf

Version 1.96
* Make compile with gcc 2.x
* root can now change passwords for other user
* timeout if client does not answer fast enough
* Sent LANG as locale to server

Version 1.95
* Make location of ldap config files compile time configurable

Version 1.94
* Add more sanity checks to prevent file corruption

Version 1.93
* Minor bug fixes to rpasswdd

Version 1.92

src/expiry.c
/* Copyright (C) 2003, 2004, 2005 Thorsten Kukuk
   Author: Thorsten Kukuk <kukuk@suse.de>

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License version 2 as
   published by the Free Software Foundation.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software Foundation,
   Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.  */


#ifdef HAVE_CONFIG_H
#include <config.h>
#endif

#include <time.h>
#include <errno.h>
#include <stdio.h>
#include <string.h>
#include <unistd.h>
#include <getopt.h>

#include "i18n.h"
#include "public.h"
#include "utf8conv.h"
#include "logging.h"
#include "error_codes.h"

#define SCALE DAY

static void
print_usage (FILE *stream, const char *program)
{
  fprintf (stream, _("Usage: %s [-f]\n"),
           program);
}

static void
print_help (const char *program)
{
  print_usage (stdout, program);
  fprintf (stdout, _("%s - check password expiration and enforce password change\n\n"), program);

  fputs (_("  -f, --force    The caller is forced to change the password\n"),
	 stdout);
  fputs (_("      --help     Give this help list\n"), stdout);
  fputs (_("  -u, --usage    Give a short usage message\n"), stdout);
  fputs (_("  -v, --version  Print program version\n"), stdout);
}

static int
c2n (char c)
{
  if (c == '.')
    return 0;
  else if (c == '/')
    return 1;
  else if (c >= '0' && c <= '9')
    return 2 + (c - '0');
  else if (c >= 'A' && c <= 'Z')
    return 12 + (c - 'A');
  else if (c >= 'a' && c <= 'z')
    return 38 + (c - 'a');
  else return -1;
}

static long
str2week (char *date)
{
  if (date == NULL || strlen (date) == 0)
    return -1;

  if (strlen (date) == 1)
    return c2n (date[0]);
  else
    return c2n (date[0]) + (c2n (date[1]) * 64);
}

static int
hp_expire (const struct passwd *pw)
{
  long min, max;
  char *age;

  age = strchr (pw->pw_passwd, ',');
  if (age == NULL)
    return 0;
  ++age;

  max = c2n (age[0]);
  if (max < 0)
    {
    error_state:
      fprintf (stderr, _("Age field for %s is wrong"),
	       utf8_to_locale (pw->pw_name));
      return -1;
    }
  ++age;

  if (age == NULL)
    goto error_state;

  min = c2n (age[0]);
  if (min < 0)
    goto error_state;
  ++age;

  if (age == NULL)
    goto error_state;

  if ((max == 0 && min == 0) ||
      ((time(0)/(SCALE*7) > str2week (age) + max) && (max >= min)))
    {
      fprintf (stdout,
	       _("Your password has expired. Choose a new password."));
      return 1;
    }

  return 0;
}

static int
expire (const struct spwd *sp)
{
  /* Print when the user has to change his password the next time ! */
  long now, remain;

  now = time (NULL) / SCALE;

  if (sp->sp_expire > 0 && now >= sp->sp_expire)
    {
      fprintf (stdout,
	       _("Your login has expired. "
		 "Contact the system administrator.\n"));
      return 3;
    }

  if (sp->sp_lstchg == 0)
    {
      fprintf (stdout,
	       _("Password changing requested. Choose a new password.\n"));
      return 1;
    }
  else if (sp->sp_lstchg > 0 && sp->sp_max >= 0 &&
           (now > sp->sp_lstchg + sp->sp_max))
    {
      if ((sp->sp_inact >= 0 &&
	   now >= sp->sp_lstchg + sp->sp_max + sp->sp_inact) ||
	  (sp->sp_max < sp->sp_min))
        {
	  fprintf (stdout, _("Your password is inactive. "
			     "Contact the system administrator.\n"));
          return 2;
        }

      fprintf (stdout, _("Your password has expired. "
			 "Choose a new password.\n"));
      return 1;
    }

  if (sp->sp_lstchg != -1 && sp->sp_max != -1 && sp->sp_warn != -1)
    if ((remain = (sp->sp_lstchg + sp->sp_max) - now) <= sp->sp_warn)
      {
	if (remain > 1)
	  fprintf (stdout, _("Your password will expire in %ld days.\n"),
		   remain);
	else if (remain == 1)
	  fprintf (stdout, _("Your password will expire tomorrow.\n"));
	else if (remain == 0)
	  fprintf (stdout, _("Your password will expire within 24 hours.\n"));
      }

  return 0;
}


int
main (int argc, char *argv[])
{
  const char *program = "expiry";
  struct passwd resultpwbuf;
  struct passwd *pw;
  struct spwd resultspbuf;
  struct spwd *sp;
  uid_t uid = getuid ();
  int force = 0, result = -1;

#ifdef ENABLE_NLS
  setlocale(LC_ALL, "");
  bindtextdomain(PACKAGE, LOCALEDIR);
  textdomain(PACKAGE);
#endif

  open_sec_log (program);

  while (1)
    {
      int c;
      int option_index = 0;
      static struct option long_options[] = {
        {"check",   no_argument, NULL, 'c' },
        {"force",   no_argument, NULL, 'f' },
        {"version", no_argument, NULL, 'v' },
        {"usage",   no_argument, NULL, 'u' },
        {"help",    no_argument, NULL, '\255' },
        {NULL,      0,           NULL, '\0'}
      };

      c = getopt_long (argc, argv, "cfvu",
                       long_options, &option_index);
      if (c == (-1))
        break;
      switch (c)
	{
	case 'c':
	  /* Dummy for compatibility with expiry from shadow suite.  */
	  break;
	case 'f':
	  ++force;
	  break;
	case '\255':
          print_help (program);
          return 0;
        case 'v':
          print_version (program, "2005");
          return 0;
        case 'u':
          print_usage (stdout, program);
	  return 0;
	default:
	  print_error (program);
	  return E_USAGE;
	}
    }

  argc -= optind;
  argv += optind;

  if (argc > 1)
    {
      fprintf (stderr, _("%s: Too many arguments.\n"), program);
      print_error (program);
      return E_USAGE;
    }
  else
    {
      int buflen = 256;
      char *buffer = malloc (buflen);

      /* Determine our own user name.  */
      while (getpwuid_r (uid, &resultpwbuf, buffer, buflen, &pw) != 0
             && errno == ERANGE)
        {
          errno = 0;
          buflen += 256;
          buffer = realloc (buffer, buflen);
        }

      if (!pw)
        {
          fprintf (stderr, _("%s: Cannot determine your user name.\n"),
                   program);
          return E_UNKNOWN_USER;
        }

      buffer = malloc (buflen);
      /* Determine our own user name.  */
      while (getspnam_r (pw->pw_name, &resultspbuf, buffer, buflen, &sp) != 0
             && errno == ERANGE)
        {
          errno = 0;
          buflen += 256;
          buffer = realloc (buffer, buflen);
        }
    }

  if (strchr (pw->pw_passwd, ',') != NULL)
    result = hp_expire (pw);
  else if (sp != NULL)
    result = expire (sp);

  if (result == -1)
    return E_FAILURE;

  if (force && result == 1)
    {
      int i;

      /* close all filehandles.  */
      for (i = 3; i < getdtablesize(); ++i)
	close (i);

      /* one single newline.  */
      fputs ("\n", stdout);

      /* drop privilegs.  */
      if (setgid (getgid ()) || setuid (uid))
        {
	  sec_log (program, MSG_DROP_PRIVILEGE_FAILED, errno, uid);
          fprintf (stderr, _("%s: Failed to drop privileges: %s\n"),
                   program, strerror (errno));
          return E_FAILURE;
        }

      execl (PASSWD_PROGRAM, PASSWD_PROGRAM, pw->pw_name, (char *)0);
      perror ("Can't execute " PASSWD_PROGRAM);
      return E_FAILURE;
    }

  return 0;
}
